org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/pom/ReImportMavenProjectsHandler.java

```diff
@@ -0,0 +1,175 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.core.internal.pom;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.eclipse.che.jdt.ls.extension.api.dto.ReImportMavenProjectsCommandParameters;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.jobs.IJobChangeEvent;
+import org.eclipse.core.runtime.jobs.IJobManager;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.jobs.JobChangeAdapter;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.managers.ProjectsManager;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
+
+/** @author Mykola Morhun */
+public class ReImportMavenProjectsHandler {
+  public static long REIMPORT_TIMEOUT = 60_000L;
+
+  private static final Gson gson =
+      new GsonBuilder()
+          .registerTypeAdapterFactory(new CollectionTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EitherTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
+          .create();
+
+  private static final PreferenceManager preferenceManager = new PreferenceManager();
+  private static final ProjectsManager projectsManager = new ProjectsManager(preferenceManager);
+  private static final IJobManager jobManager = Job.getJobManager();
+  private static final JobChangedListener jobChangedListener = new JobChangedListener();
+  private static final Lock lock = new ReentrantLock();
+
+  private static CountDownLatch jobsToBeFinished;
+  // name -> path
+  private static Map<String, String> projectsToBeUpdated = new ConcurrentHashMap<>();
+  // set of paths to updates projects; path -> ""
+  private static Map<String, String> updatedProjectsPaths = new ConcurrentHashMap<>();
+
+  /**
+   * Updates given maven projects.
+   *
+   * @param arguments contains ReImportMavenProjectsCommandParameters in the first element
+   * @param progressMonitor progress monitor
+   * @return URIs of updated projects
+   */
+  public static List<String> reImportMavenProjects(
+      List<Object> arguments, IProgressMonitor progressMonitor) {
+
+    if (lock.tryLock()) {
```

Comment:
would it be simpler to make this method synchronized?

Reply:
No, because we have to skip queries if one in progress.

Reply:
Even if we want to wait, using a lock would be better, because it is
1. does not lock the whole object
2. Can be interrupted
3. Can have a timeout.

Reply:
ok